samples/bluetooth/direct_adv/src/main.c

```diff
@@ -0,0 +1,169 @@
+ * Copyright (c) 2015-2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/reboot.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/bluetooth/gatt.h>
+
+/* Custom Service Variables */
+#define BT_UUID_CUSTOM_SERVICE_VAL \
+	BT_UUID_128_ENCODE(0x12345678, 0x1234, 0x5678, 0x1234, 0x56789abcdef0)
+
+static struct bt_uuid_128 primary_service_uuid = BT_UUID_INIT_128(
+	BT_UUID_CUSTOM_SERVICE_VAL);
+
+static struct bt_uuid_128 read_characteristic_uuid = BT_UUID_INIT_128(
+	BT_UUID_128_ENCODE(0x12345678, 0x1234, 0x5678, 0x1234, 0x56789abcdef1));
+
+static struct bt_uuid_128 write_characteristic_uuid = BT_UUID_INIT_128(
+	BT_UUID_128_ENCODE(0x12345678, 0x1234, 0x5678, 0x1234, 0x56789abcdef2));
+
+static int signed_value;
+static struct bt_le_adv_param adv_param;
+static bt_addr_le_t bond_addr;
+
+static ssize_t read_signed(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			   void *buf, uint16_t len, uint16_t offset)
+{
+	int *value = &signed_value;
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, value,
+				 sizeof(signed_value));
+}
+
+static ssize_t write_signed(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			    const void *buf, uint16_t len, uint16_t offset,
+			    uint8_t flags)
+{
+	int *value = &signed_value;
+
+	if (offset + len > sizeof(signed_value)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	memcpy(value + offset, buf, len);
+
+	return len;
+}
+
+/* Vendor Primary Service Declaration */
+BT_GATT_SERVICE_DEFINE(primary_service,
+	BT_GATT_PRIMARY_SERVICE(&primary_service_uuid),
+	BT_GATT_CHARACTERISTIC(&read_characteristic_uuid.uuid,
+			       BT_GATT_CHRC_READ,
+			       BT_GATT_PERM_READ,
+			       read_signed, NULL, NULL),
+	BT_GATT_CHARACTERISTIC(&write_characteristic_uuid.uuid,
+			       BT_GATT_CHRC_WRITE,
+			       BT_GATT_PERM_WRITE_ENCRYPT,
+			       NULL, write_signed, NULL),
+);
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, BT_UUID_CUSTOM_SERVICE_VAL),
+};
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err) {
+		printk("Connection failed (err 0x%02x)\n", err);
+	} else {
+		printk("Connected\n");
+	}
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	printk("Disconnected (reason 0x%02x)\n", reason);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected
+};
+
+static void bond_loop(const struct bt_bond_info *info, void *data)
+{
+	uint8_t *counter = data;
+
+	bt_addr_le_copy(&bond_addr, &info->addr);
+	(*counter)++;
```

Comment:
why do we need the counter? I don't see it used anywhere, it just looks like you're just assigning the directed-adv address to the last-bonded device.

Reply:
Counter is used at L122, to check if at least one device is bonded. Should I change it to boolean variable? Or maybe there is elegant way to check whether `bond_addr` is empty, so I can start direct advertisting if it is not empty?

Reply:
yes there is :)
```c
bt_addr_le_copy(&bond_addr, BT_ADDR_LE_ANY);

bt_foreach_bond(BT_ID_DEFAULT, bond_loop, &bond_count);

/* Addresses are equal if compare returns 0. In that case,
 * That means the address has not yet been set.
 */
if (bt_addr_le_cmp(&bond_addr, BT_ADDR_LE_ANY) != 0) {
... }
```

Reply:
I went with BT_ADDR_LE_NONE, as I think it better represents the "empty" state